src/Control/Effect/Fresh.hs

```diff
@@ -29,29 +34,36 @@ instance Effect Fresh where
 --
 --   prop> run (runFresh (replicateM n fresh)) == nub (run (runFresh (replicateM n fresh)))
 fresh :: (Member Fresh sig, Carrier sig m) => m Int
-fresh = send (Fresh ret)
+fresh = send (Fresh pure)
 
 -- | Reset the fresh counter after running a computation.
 --
 --   prop> run (runFresh (resetFresh (replicateM m fresh) *> replicateM n fresh)) == run (runFresh (replicateM n fresh))
 resetFresh :: (Member Fresh sig, Carrier sig m) => m a -> m a
-resetFresh m = send (Reset m ret)
+resetFresh m = send (Reset m pure)
 
 
 -- | Run a 'Fresh' effect counting up from 0.
 --
 --   prop> run (runFresh (replicateM n fresh)) == [0..pred n]
 --   prop> run (runFresh (replicateM n fresh *> pure b)) == b
-runFresh :: (Carrier sig m, Effect sig, Monad m) => Eff (FreshC m) a -> m a
-runFresh = fmap snd . flip runFreshC 0 . interpret
+runFresh :: Functor m => FreshC m a -> m a
+runFresh = evalState 0 . runFreshC
 
-newtype FreshC m a = FreshC { runFreshC :: Int -> m (Int, a) }
+newtype FreshC m a = FreshC { runFreshC :: StateC Int m a }
+  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadIO, MonadPlus, MonadTrans)
 
-instance (Carrier sig m, Effect sig, Monad m) => Carrier (Fresh :+: sig) (FreshC m) where
-  ret a = FreshC (\ i -> ret (i, a))
-  eff op = FreshC (\ i -> handleSum (eff . handleState i runFreshC) (\case
-    Fresh   k -> runFreshC (k i) (succ i)
-    Reset m k -> runFreshC m i >>= flip runFreshC i . k . snd) op)
+instance (Carrier sig m, Effect sig) => Carrier (Fresh :+: sig) (FreshC m) where
+  eff (L (Fresh   k)) = FreshC $ do
+    i <- get
+    put (succ i)
+    runFreshC (k i)
```

Comment:
Defining `FreshC` in terms of `StateC` means that we can use the `State` effect directly in the `Carrier` instance, clarifying the definition substantially.